resource-registry/src/Control/ResourceRegistry.hs

```diff
@@ -460,6 +464,12 @@ allocKey = unlessClosed $ do
   modify $ \st -> st{registryNextKey = succ nextKey}
   return nextKey
 
+allocNKeys :: Int -> State (RegistryState m) (Either PrettyCallStack [ResourceId])
```

Comment:
I think the code duplication between `allocNKeys` and `allocKey` is very confusing, as it's syntactically almost the same code. Could you refactor these two functions so that `allocNKeys` is expressed in terms of `allocKey`?